src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -380,5 +405,26 @@ private static string GetDisplayGroup(FieldInfo field)
 
             return string.Empty;
         }
+
+        // Internal for testing
+        internal static bool IsNonNullable(IEnumerable<object> attributes)
+        {
+            // [Nullable] is compiler synthesized, comparing by name. 
+            var nullableAtttribute = attributes
+                .Where(a => string.Equals(a.GetType().FullName, NullableAttributeFullTypeName))
+                .FirstOrDefault();
+
+            // See: https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md#annotations
+            if (nullableAtttribute != null &&
+                nullableAtttribute.GetType().GetField(NullableFlagsFieldName) is FieldInfo field &&
+                field.GetValue(nullableAtttribute) is byte[] flags
+                && flags.Length >= 0
+                && flags[0] == 1)
```

Comment:
Does this have trouble with collection types? 

> [Nullable(new[] { 1, 2 })] string[] NotNullMaybeNull; // string?[]!

if not, we should add some comments explaining how this works when more than one value is present in the `NullableAttribute` byte array

Reply:
Yeah, I hadn't thought about this case, but it's worth discussion. 

Note that we don't have anything like this today so this would be a one-off thing or something built into the model binders for collections. We don't support you placing `[Required]` on a type parameter.

Reply:
There's actually another problem here.

Consider a case like:

```C#
    class Program
    {
        public string Foo { get; set; } = default!;

        public Container<string?> Bar { get; set; } = default!;
    }

    class Container<T>
    {
        public T Item { get; set; } = default!;
    }
```

ModelMetadata is an API you can use to get the metadata for a property, parameter or type.  We're primarily interested in the property case, where you pass in the container type and property info. However, with nnrt, this isn't enough information.  In this example for property `Item` the only place where the metadata reflect the nnrt information is on property `Bar`. 

You need to know the property, the containing type and now the *containing property* in order to create a correct model metadata. Basically with generics in the picture, our current architecture can't support this usage. Put another way, nnrt + generics is *just* type erasure :(